File: 00238. Product of Array Except Self/00238.c
int* productExceptSelf(int* nums, int numsSize, int* returnSize){
    int i, product, zero_count;
    int *ret = (int*) malloc(sizeof(int) * numsSize);
    
    for (i=0, zero_count=0, product=1; i<numsSize; ++i) {
        if (nums[i] == 0) {
            zero_count++;
        } else {
            product *= nums[i];
        }
    }
    
    for (i=0; i<numsSize; ++i) {
        if (nums[i] == 0) {
            // There is at least one more zero in nums[]
            if (zero_count > 1) {
                ret[i] = 0;
            } else {
                ret[i] = product;
            }
        } else {
            // There is at least one zero in nums[]
            if (zero_count != 0) {
                ret[i] = 0;
            } else {
                ret[i] = product / nums[i];
            }
        }
    }
    
    *returnSize = numsSize;
    return ret;
}

File: LeetCode/00238. Product of Array Except Self/00238.go
// https://leetcode.com/problems/product-of-array-except-self

func productExceptSelf(nums []int) []int {
    size := len(nums)
    ans := make([]int, size, size)

    zero, zeroIdx, product := 0, 0, 1
    for i := 0; i < size; i++ {
        if nums[i] == 0 {
            zero, zeroIdx = zero+1, i
            if zero == 2 {
                break
            }

            continue
        }

        product *= nums[i]
    }

    if zero > 1 {
        return make([]int, size, size)
    } else if zero == 1 {
        ans = make([]int, size, size)
        ans[zeroIdx] = product
    } else {
        for i := 0; i < size; i++ {
            ans[i] = product / nums[i]
        }
    }

    return ans
}
